### ISSUE TITLE
Improve sct_run_batch cluster awareness

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/spinalcordtoolbox#2806]
Currently, `sct_run_batch` works with OpenMP, not MPI.  In clusters that can accommodate MPI, it would be nice to have that capabilities. In case we will implement it, we could re-use some of the code from an old (now deprecated https://github.com/neuropoly/spinalcordtoolbox/pull/2759) SCT function [`sct_pipeline`](https://github.com/neuropoly/spinalcordtoolbox/blob/09a47b2bed7f13d04bd01a962948883f6ccb7f38/scripts/sct_pipeline.py#L55).

**Comment by jcohenadad** [comment]
suggestions from compute canada staff:
~~~
  As my colleague mentioned, the Python multiprocessing package only allows you to run in parallel within a single node and its CPU cores (from 32 to 48 cores, depending on the cluster and node). As for whether Niagara is the right cluster for this, perhaps not given that such a job is indifferent to the quality of the network connecting the compute nodes - you might be better running on one of the general purpose clusters like Béluga, Graham and Cedar. If you do wish to be able to use more than one node for your computations, another approach which might be better suited to your workflow is GNU Parallel,
https://docs.computecanada.ca/wiki/GNU_Parallel
or GLOST,
https://docs.computecanada.ca/wiki/GLOST
~~~

**Comment by cfhammill** [comment]
I was worried that this was the case, but had talked myself out of it. I think in order to get `sct_run_batch` to be more cluster aware would involve a substantial rewrite. @bcdarwin has worked on a cluster aware pipelining framework called pydpiper for a long time https://github.com/Mouse-Imaging-Centre/pydpiper, although he might suggest using an alternative framework depending on what the needs of the project are. Nipype supports slurm for instance https://nipype.readthedocs.io/en/latest/api/generated/nipype.pipeline.plugins.slurmgraph.html, and refactoring `sct_run_batch` to use nipype might be the path of least resistance.

**Comment by jcohenadad** [comment]
@cfhammill i vaguely remember that we used to run `sct_pipeline` in MPI mode, see [here](https://github.com/neuropoly/spinalcordtoolbox/blob/09a47b2bed7f13d04bd01a962948883f6ccb7f38/scripts/sct_pipeline.py#L55). It doesn't seem like there is a lot of change required in the current code, but maybe i misunderstand the concept of how MPI should be used.

**Comment by cfhammill** [comment]
I think the code in `sct_pipeline` is sufficient to get MPI to run, but as I mentioned in the slack thread I don't think this will be smart enough to avoid starting too many jobs on a single node. We may be able to fix this in the slurm configuration for the job though. I'll work on adapting the current code, it should be relatively simple although there may be some subtleties switching from `multiprocessing` to `futures`. Then I can try to figure out the necessary configuration for niagara. 

Another alternative I wasn't aware of, very recently @bcdarwin has had some success converting our code to use [makeflow](http://ccl.cse.nd.edu/software/makeflow/), this transparently handles a variety of different parallelization strategies (remote multiprocessing with torque/slurm, local single/multiprocessing with workqueue) this is probably more robust and handles restarting so that the pipeline is aware if steps don't need to be rerun. I think converting `sct_run_batch` to makeflow would be pretty straight forward, but it does incur an extra depency (although available through pypi). I suspect makeflow wouldn't be trivial to work effectively on niagara (1 slurm job per node, plus local parallelization on the nodes), but it would work like a charm on graham.

**Comment by cfhammill** [comment]
PR open adding this functionality, testing on niagara to make sure it works.

**Comment by cfhammill** [comment]
I'm having trouble getting time on niagara for an interactive job to test, will try again soon.

**Comment by cfhammill** [comment]
After fighting with MPI on niagara for a bit today it looks like the gcc compatible openmpi they have installed has a bug (intel compilers are more completely supported by Compute Canada as far as I can tell). I will report it and wait for an update.

**Comment by cfhammill** [comment]
After fighting with MPI on niagara as described in #2827 and determining that it would be a substantial effort, I think we should potentially close this issue for now. @jcohenadad do you agree? Or perhaps we can rename this, improve sct_run_batch cluster awareness, to remain open to alternative solutions to cluster based parallel computing.

**Comment by jcohenadad** [comment]
@PaulBautin has put together a [very useful script](https://github.com/sct-pipeline/csa-atrophy/blob/master/run_all.py) to break down a dataset and launch multiple SLURM jobs on a cluster.

@cfhammill i'm thinking we could include that script as part of SCT? or maybe as a separate repository (could be its own package), given the specific usage. What do you think?

**Comment by cfhammill** [comment]
I think it could potentially be generalized and be included in SCT. In it's present state it's very specific to a user in your compute canada group running on a SLURM cluster. Earlier we had discussed `sct_run_batch` being primarily for internal use (hence some of the design choices regarding email), so this may not be a problem. 

It would be more satisfying to me if we had cluster awareness baked in to `sct_run_batch`, so that it could be used under a variety of cluster computing setups, but maybe this covers all of our use cases for now.
